tools/gz_msgs_generate_factory.py

```diff
@@ -0,0 +1,167 @@
+#!/usr/bin/env python3
+#
+# Copyright (C) 2023 Open Source Robotics Foundation
+#
+# Licensed under the Apache License, Version 2.0 (the "License")
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#       http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import argparse
+import os
+import pathlib
+import sys
+
+# Create <gz/msgs/MessageTypes.hh>
+cc_header = """/*
+ * Copyright (C) 2023 Open Source Robotics Foundation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+*/
+
+/* This file was automatically generated.
+ * Do not edit this directly
+ */
+
+#ifndef GZ_MSGS_MESSAGE_TYPES_HH_
+#define GZ_MSGS_MESSAGE_TYPES_HH_
+{gz_msgs_headers}
+
+namespace {namespace} {{
+int RegisterAll();
+}}
+#endif"""
+
+# Create factory registration bits
+cc_source = """/*
+ * Copyright (C) 2023 Open Source Robotics Foundation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+*/
+
+/* This file was automatically generated.
+ * Do not edit this directly
+ */
+
+#include "gz/msgs/Factory.hh"
+#include "gz/msgs/MessageFactory.hh"
+#include "{package_path}/MessageTypes.hh"
+
+#include <map>
```

Comment:
Creating a `std::map` global variable violates the static storage rules in our 
style guide. I recommend using the codegen here to create a `std::array` and 
call `Factory::Register` while looping on the array. A new map 
is being created in `MessageFactory` anyway, so it would actually be more 
performant.